internal/store/bundle.go

```diff
@@ -128,7 +144,36 @@ func (b *bundleStore) Remove(ref reference.Reference) error {
 		return errors.New("no such image " + reference.FamiliarString(ref))
 	}
 	b.refsMap.removeRef(ref)
-	return os.RemoveAll(path)
+
+	if err := os.RemoveAll(path); err != nil {
+		return nil
+	}
+	return cleanupParentTree(path)
+}
+
+func cleanupParentTree(path string) error {
+	for {
+		path = filepath.Dir(path)
+		if empty, err := isEmpty(path); err != nil || !empty {
+			return err
+		}
+		if err := os.RemoveAll(path); err != nil {
+			return nil
+		}
+	}
+}
+
+func isEmpty(path string) (bool, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, err
+	}
+	if _, err = f.Readdir(1); err == io.EOF {
+		// dir is empty
+		return true, nil
+	}
+	defer f.Close()
```

Comment:
Shouldn't we defer before `Readdir` ? Otherwise we can return without closing if `err == io.EOF` 🤔